LEGO1/mxcompositepresenter.cpp

```diff
@@ -43,34 +50,106 @@ MxCompositePresenter::~MxCompositePresenter()
 	NotificationManager()->Unregister(this);
 }
 
-// STUB: LEGO1 0x100b6410
-MxResult MxCompositePresenter::StartAction(MxStreamController*, MxDSAction*)
+// FUNCTION: LEGO1 0x100b6410
+MxResult MxCompositePresenter::StartAction(MxStreamController* p_controller, MxDSAction* p_action)
 {
-	// TODO
-	return SUCCESS;
+	MxAutoLocker lock(&m_criticalSection);
+
+	MxResult result = FAILURE;
+	MxDSActionList* actions = ((MxDSMultiAction*) p_action)->GetActionList();
+	MxObjectFactory* factory = ObjectFactory();
+	MxDSActionListCursor cursor(actions);
+	MxDSAction* action;
+
+	if (MxPresenter::StartAction(p_controller, p_action) == SUCCESS) {
+		// The usual cursor.Next() loop doesn't match here, even though
+		// the logic is the same. It does match when "deconstructed" into
+		// the following Head(), Current() and NextFragment() calls,
+		// but this seems unlikely to be the original code.
+		// The alpha debug build also uses Next().
+		cursor.Head();
+		while (cursor.Current(action)) {
+			cursor.NextFragment();
+
+			MxBool success = FALSE;
+
+			action->CopyFlags(m_action->GetFlags());
+
+			const char* presenterName = PresenterNameDispatch(*action);
+			MxPresenter* presenter = (MxPresenter*) factory->Create(presenterName);
+
+			if (presenter && presenter->AddToManager() == SUCCESS) {
+				presenter->SetCompositePresenter(this);
+				if (presenter->StartAction(p_controller, action) == SUCCESS)
+					success = TRUE;
+			}
+
+			if (success) {
+				action->SetOrigin(this);
+				m_list.push_back(presenter);
+			}
+			else if (presenter)
+				delete presenter;
+		}
+
+		result = SUCCESS;
+	}
+
+	return result;
 }
 
-// STUB: LEGO1 0x100b65e0
+// FUNCTION: LEGO1 0x100b65e0
 void MxCompositePresenter::EndAction()
 {
-	// TODO
+	MxAutoLocker lock(&m_criticalSection);
+
+	if (!m_action)
+		return;
+
+	((MxDSMultiAction*) m_action)->GetActionList()->DeleteAll(FALSE);
+
+	while (!m_list.empty()) {
+		MxPresenter* presenter = m_list.front();
+		m_list.pop_front();
+		presenter->SetCompositePresenter(NULL);
+		presenter->EndAction();
+	}
+
+	MxDSAction* action = m_action;
+	MxPresenter::EndAction();
+
+	if (action && action->GetOrigin()) {
+		NotificationManager()->Send(
+			action->GetOrigin(),
+			&MxEndActionNotificationParam(c_notificationEndAction, this, action, FALSE)
+		);
+	}
 }
 
-// STUB: LEGO1 0x100b6760
+// FUNCTION: LEGO1 0x100b6760
 MxLong MxCompositePresenter::Notify(MxParam& p)
```

Comment:
I might be wrong, and I feel like we've had this discussion before but given the 1996 source are these 'MxS32'?

Reply:
The function is exported with `long` as a return type, so it must be `MxLong`.